File: yggdrasil/examples/maxMsg/src/maxMsgCli.cpp
#include "YggInterface.hpp"
#include <stdio.h>


void rand_str(char *dest, size_t length) {
  char charset[] = "0123456789"
    "abcdefghijklmnopqrstuvwxyz"
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  while (length-- > 1) {
    size_t index = (size_t)((double) rand() / RAND_MAX * (sizeof charset - 1));
    *dest++ = charset[index];
  }
  *dest = '\0';
}


int main(int argc, char *argv[]) {

    //char output[YGG_MSG_BUF];
    //char input[YGG_MSG_BUF];
    size_t msg_size_output = YGG_MSG_BUF;
    size_t msg_size_input = YGG_MSG_BUF;
    char *output = (char*)malloc(msg_size_output);
    char *input = (char*)malloc(msg_size_input);
  
    printf("maxMsgCli(CPP): Hello message size is %d.\n", (int)msg_size_output);
    
    // Create a max message, send/recv and verify
    YggRpcClient rpc("maxMsgSrv_maxMsgCli", "%s", "%s");
    
    // Create a max message
    rand_str(output, msg_size_output);
    printf("maxMsgCli(CPP): sending %.10s...\n", output);
    
    // Call RPC server
    if (rpc.callRealloc(4, output, msg_size_output-1,
			&input, &msg_size_input) < 0) {
      printf("maxMsgCli(CPP): RPC ERROR\n");
      free(output);
      free(input);
      return -1;
    }
    printf("maxMsgCli(CPP): received %lu bytes: %.10s...\n", msg_size_input, input);

    // Check to see if response matches
    if (strncmp(output, input, msg_size_output-1)) {
        printf("maxMsgCli(CPP): ERROR: input/output do not match\n");
        free(output);
        free(input);
	    return -1;
    } else {
        printf("maxMsgCli(CPP): CONFIRM\n");
    }

    // All done, say goodbye
    printf("maxMsgCli(CPP): Goodbye!\n");

    free(output);
    free(input);
    return 0;
}

File: yggdrasil/examples/maxMsg/src/maxMsgCli.f90
program main
  use fygg

  integer(kind=c_size_t) :: msg_size_output = YGG_MSG_BUF
  integer(kind=c_size_t) :: msg_size_input = YGG_MSG_BUF
  character, dimension(YGG_MSG_BUF) :: output
  type(yggchar_r) :: input
  type(yggcomm) :: rpc
  logical :: ret

  print *, "maxMsgCli(F): Hello message size is ", msg_size_output

  ! Create a max message, send/recv and verify
  rpc = ygg_rpc_client("maxMsgSrv_maxMsgCli", "%s", "%s")

  ! Create a max message
  call rand_str(output, msg_size_output)
  print *, "maxMsgCli(F): sending ", output(1:10), "..."

  ! Call RPC server
  msg_size_input = size(input%x)
  ret = ygg_rpc_call_realloc(rpc, &
       [yggarg(output), yggarg(msg_size_output)], &
       [yggarg(input), yggarg(msg_size_input)])
  if (.not.ret) then
     print *, "maxMsgCli(F): RPC ERROR"
     stop 1
  end if
  print *, "maxMsgCli(F): received ", msg_size_input, " bytes: ", &
       input%x(1:10), "..."

  ! Check to see if response matches
  if (.not.all(output.eq.(input%x))) then
     print *, "maxMsgCli(F): ERROR: input/output do not match"
     stop 1
  else
     print *, "maxMsgCli(F): CONFIRM"
  end if

  ! All done, free and say goodbye
  print *, "maxMsgCli(F): Goodbye!"

end program main

subroutine rand_str(x, x_siz)
  use fygg
  character, dimension(*) :: x
  real :: x_rand
  integer(kind=c_size_t) :: x_siz, i, index
  character(len=62) :: charset
  charset = "0123456789&
       &abcdefghijklmnopqrstuvwxyz&
       &ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  do i = 1, x_siz
     call random_number(x_rand)
     index = int(x_rand * (len(charset) - 1), c_size_t) + 1
     x(i) = charset(index:index)
  end do
  ! do i = (x_siz+1), size(x)
  !    x(i) = ' '
  ! end do
end subroutine rand_str

